Sources/SourceControl/GitRepository.swift

```diff
@@ -6,44 +6,73 @@
 
  See http://swift.org/LICENSE.txt for license information
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
-*/
+ */
 
 import Basics
-import TSCBasic
 import Dispatch
+import TSCBasic
 import TSCUtility
 
-/// A `git` repository provider.
-public struct GitRepositoryProvider: RepositoryProvider {
+// MARK: - GitShellHelper
 
+/// Helper for shelling out to `git`
+private struct GitShellHelper {
     /// Reference to process set, if installed.
     private let processSet: ProcessSet?
 
-    public init(processSet: ProcessSet? = nil) {
+    init(processSet: ProcessSet? = nil) {
         self.processSet = processSet
     }
 
     /// Private function to invoke the Git tool with its default environment and given set of arguments.  The specified
     /// failure message is used only in case of error.  This function waits for the invocation to finish and returns the
     /// output as a string.
-    @discardableResult
-    private func callGit(_ args: String..., environment: [String: String] = Git.environment, failureMessage: String = "", repository: RepositorySpecifier) throws -> String {
+    func run(_ args: [String], environment: [String: String] = Git.environment) throws -> String
+    {
         let process = Process(arguments: [Git.tool] + args, environment: environment, outputRedirection: .collect)
         let result: ProcessResult
         do {
-            try processSet?.add(process)
+            try self.processSet?.add(process)
             try process.launch()
             result = try process.waitUntilExit()
-        }
-        catch {
+            guard result.exitStatus == .terminated(code: 0) else {
+                throw GitShellError(result: result)
+            }
+            return try result.utf8Output().spm_chomp()
```

Comment:
For a later PR, we might want to wrap any error thrown by this `try` in an error that provides the context for the Git operation that was attempted (I think it will just show up as a UTF-8 decoding error otherwise).  In fact it might be better to use a more lenient decoder here, but that might require a change to `ProcessResult`.

Reply:
So this is fine for this PR, I think.